### PR TITLE
CatVariants can mis-sort input VCFs without -assumeSorted

### PR DESCRIPTION
**Pull request by johnwallace123**
When using CatVariants, VCF files were being sorted solely on the base
pair position of the first record, ignoring the chromosome.  This can
become problematic when merging files from different chromosomes,
espeically if you have multiple VCFs per chromosome.

As an example, assume the following 3 lines are all in separate files:
1       10
1       100
2       20

The merged VCF from CatVariants (without -assumeSorted) would read:
1       10
2       20
1       100

This has the potential to break tools that expect chromosomes to be
contiguous within a VCF file.

This commit changes the comparator from one of Pair\<Integer, File\> to
one of Pair\<VariantContext, File\>.  We construct a
VariantContextComparator from the provided reference, which will sort
the first record by chromosome and position properly.  Additionally, if
-assumeSorted is given, we simply use a null VariantContext as the first
record, which will all be equal (as all will be null)

### CODE REVIEW

**Comment by vdauwera** [comment]
@johnwallace123 Sorry it's taken us a while to get to this! We're processing it now.

File: public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java [hunk 1]
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
...
-            if (startPositionP1  == startPositionP2)
-                return 0;
-            return startPositionP1 < startPositionP2 ? -1 : 1 ;
+        public int compare(Pair<VariantContext,File> p1, Pair<VariantContext,File> p2) {
+            VariantContext startPositionP1 = p1.getFirst();
+            VariantContext startPositionP2 = p2.getFirst();
+            return comp.compare(startPositionP1, startPositionP2);
+            //if (startPositionP1  == startPositionP2)
+            //    return 0;
+            //return startPositionP1 < startPositionP2 ? -1 : 1 ;
```
Comment (ronlevine) [inline public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java]: Remove commented out lines

Comment (ronlevine) [inline public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java]: Add a test to CatVariantsIntegrationTes for checking the sorting logic/comparator.

File: public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java [hunk 2]
```diff
@@ -257,9 +259,9 @@ protected int execute() throws Exception {
                     continue;
                 }
                 VariantContext vc = it.next();
-                int firstPosition = vc.getStart();
+                //int firstPosition = vc.getStart();
```
Comment (ronlevine) [inline public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java]: Remove commented out line.

File: public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java [hunk 3]
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
         }
     }
 
-    private static class PositionComparator implements Comparator<Pair<Integer,File>> {
+    private static class PositionComparator implements Comparator<Pair<VariantContext,File>> {
+    	
+    	VariantContextComparator comp;
+    	
+    	public PositionComparator(SAMSequenceDictionary dict){
```
Comment (ronlevine) [inline public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java]: make argument "final".

File: public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java [hunk 4]
```diff
@@ -318,15 +320,22 @@ public static void main(String[] args){
...
+    	}
 
         @Override
-        public int compare(Pair<Integer,File> p1, Pair<Integer,File> p2) {
-            int startPositionP1 = p1.getFirst();
-            int startPositionP2 = p2.getFirst();
-            if (startPositionP1  == startPositionP2)
-                return 0;
-            return startPositionP1 < startPositionP2 ? -1 : 1 ;
+        public int compare(Pair<VariantContext,File> p1, Pair<VariantContext,File> p2) {
```
Comment (ronlevine) [inline public/gatk-tools-public/src/main/java/org/broadinstitute/gatk/tools/CatVariants.java]: Make arguments "final".

**Comment by johnwallace123** [comment]
@ronlevine I believe I've added all the changes that you requested.  If you approve, I'll rebase into a single commit, otherwise I'll keep working :).

On a related note, I wasn't able to actually get any tests to run using "mvn verify" or "mvn integration-test"; it would always say "Tests are skipped".  I verified by adding an Assert.fail() into the test that I added.  Is there something that I'm doing wrong?

Thanks!

**Comment by vdauwera** [comment]
@johnwallace123 Not your fault, the test framework isn't available in the public source. We know that sucks -- will be done differently in GATK4. Apologies if you spent time trying to get the tests to work.

**Comment by ronlevine** [comment]
@johnwallace123 Don't worry about the tests. I already cherry-picked your changes, put them on a branch in a Broad private repo and created an integration test for your code. The command to run the integration test is:
./ant-bridge.sh integrationtest -Dsingle=CatVariantsIntegrationTest
But it will not run successfully for you because it needs to access Broad internal data

**Comment by johnwallace123** [comment]
@vdauwera @ronlevine:  Thanks for confirming that I wasn't going crazy!  If you need anything else from me, please let me know.

**Pull request closed without merging**